### ISSUE TITLE
enable-unit and enable-integration test report with 1 failed out of 128

### ISSUE THREAD

**Issue by rzhai** [tpm2-software/tpm2-tss#1227]
ubuntu 18.04
tpm simulator ibmtpm1332

build:
./bootstrap -I /usr/share/gnulib/m4
./configure \
--enable-unit \
--enable-integration \
--with-udevrulesdir=/etc/udev/rules.d/
make -j$(nproc) check

Testsuite summary for tpm2-tss 2.1.0
============================================================================
# TOTAL: 128
# PASS:  125
# SKIP:  2
# XFAIL: 0
# FAIL:  0
# XPASS: 0
# ERROR: 1
============================================================================
See ./test-suite.log

cat ./test-suite.log 
======================================
   tpm2-tss 2.1.0: ./test-suite.log
======================================

# TOTAL: 128
# PASS:  125
# SKIP:  2
# XFAIL: 0
# FAIL:  0
# XPASS: 0
# ERROR: 1

.. contents:: :depth: 2

ERROR: test/integration/esys-pp-commands
========================================

Starting simulator on port 41540
successfully started daemon: tpm_server with PID: 25784
/home/ubuntu/tpm2-tss
simulator PID: 25784
Simulator with PID 25784 bound to port 41540 and  41541 successfully.
WARNING:tcti:src/util/io.c:248:socket_connect() Failed to connect to host 127.0.0.1, port 41540: errno 111: Connection refused 
Failed to initialize tcti context: 0xa000a
TPM_StartUp failed
ERROR test/integration/esys-pp-commands.int (exit status: 99)

SKIP: test/integration/esys-field-upgrade
=========================================

Starting simulator on port 50926
successfully started daemon: tpm_server with PID: 26089
/home/ubuntu/tpm2-tss
simulator PID: 26089
Simulator with PID 26089 bound to port 50926 and  50927 successfully.
Execute the test script
WARNING:esys:src/tss2-esys/api/Esys_FieldUpgradeData.c:301:Esys_FieldUpgradeData_Finish() Received TPM Error 
ERROR:esys:src/tss2-esys/api/Esys_FieldUpgradeData.c:108:Esys_FieldUpgradeData() Esys Finish ErrorCode (0x00000143) 
Script returned 77
SKIP test/integration/esys-field-upgrade.int (exit status: 77)

SKIP: test/integration/esys-firmware-read
=========================================

Starting simulator on port 15986
successfully started daemon: tpm_server with PID: 26380
/home/ubuntu/tpm2-tss
simulator PID: 26380
Simulator with PID 26380 bound to port 15986 and  15987 successfully.
Execute the test script
WARNING:esys:src/tss2-esys/api/Esys_FirmwareRead.c:284:Esys_FirmwareRead_Finish() Received TPM Error 
ERROR:esys:src/tss2-esys/api/Esys_FirmwareRead.c:100:Esys_FirmwareRead() Esys Finish ErrorCode (0x00000143) 
Script returned 77
SKIP test/integration/esys-firmware-read.int (exit status: 77)

**Comment by tstruk** [comment]
We recommend using ibmtpm974

**Comment by AndreasFuchsSIT** [comment]
The firmware-update stuff is of course skipped, because it's not implemented on the simulator.

Regarding the other error "connection refused", I wonder, how many threads `nproc` gives to you. I have it running with 40 on the server I think.
Can you post `nproc` and also test `make -j1 check` ?

**Comment by rzhai** [comment]
@tstruk One of reasons I can't use ibmtpm974 is that in ubuntu 18.04, make has error.
````
CryptRand.c: In function ‘CryptRandomStir’:
CompilerDependencies.h:145:34: error: ‘dfResult’ is used uninitialized in this function [-Werror=uninitialized]
 #   define  NOT_REFERENCED(x) (x = x)
                                  ^
cc1: all warnings being treated as errors
makefile:79: recipe for target 'CryptRand.o' failed
make: *** [CryptRand.o] Error 1
````
But we don't need discuss how to solve ibm error here.

**Comment by rzhai** [comment]
@AndreasFuchsSIT 
$ echo $(nproc)
4

$ make -j1 check
```
Testsuite summary for tpm2-tss 2.1.0
============================================================================
# TOTAL: 128
# PASS:  126
# SKIP:  2
# XFAIL: 0
# FAIL:  0
# XPASS: 0
# ERROR: 0
============================================================================
```

**Comment by AndreasFuchsSIT** [comment]
Ok... Can you reproduce the failure with the make -j4 check ?

The checks have some probabilistic to them when run in parallel, cause they are racing for tcp ports.
Maybe you hit the 1:10000 case where it gets a hickup. If however this behavior is consistent, we should dive into it a little more...

**Comment by diabonas** [comment]
I occasionally hit this error as well, so if there is a way to make this more robust, it would be much appreciated :) It is however always a transient error on a different test, simply running the whole test suite a second time if it occurs causes all tests to pass.

**Comment by AndreasFuchsSIT** [comment]
The racy ugly code is here: https://github.com/tpm2-software/tpm2-tss/blob/master/script/int-log-compiler.sh#L190
I assume that the problem stems from the long wait from loading tpm_server until tpm_server actually binds the port.

If anybody wants to get something nicer done, please take a stab at it...
- If  we use some function to take $PID % 1000 + 30000 which is however also not unique if too many processes spawn in parallel...
- We could add lock-files for ports each simulator wants to allocate ? This could even make make check faster without the backoff sleeps... Maybe using liblockfile-bin ?
- In either case, we still have to check for other applications binding to ports in our range (even during execution).

First step could be to make this XFAIL instead of ERROR, because its a harness-problem, not an actual failed test...

**Comment by tstruk** [comment]
Or we can make it more deterministic and statically assign a port number to a test. We have 124 integ tests now + tpm_client, that 125 * 2 ports. This should make it a little faster as well.

**Comment by AndreasFuchsSIT** [comment]
but how do we give them an enumeration without having to statically assign them, because I don't want to be doing that... ;-)
I guess I'd like to go for $PID%1000 with lock files from liblockfile-bin...

**Comment by AndreasFuchsSIT** [comment]
Though the immediate problem was solved by #1234 I still think the improvement for more reliability as described above would be nice. Thus, keeping it open...

**Comment by tstruk** [comment]
Couldn't reproduce this issue with the current code base. Please reopen if it still affects you.
